<div align="center">

**PRACTICAL REPORT**

**ALGORITHMS AND PROGRAMMING**

**“Module 2: Data Types, Variables, and I/O”**

**Compiled By:**

**Laiba Shamshad**

**2611528002**

**Lecturer:**

**Dr. Ir WAHYUDI, ST., MT.**

**PRACTICAL ASSISTENT:**

**Sherly Sukmadira**

<br>

 <img src="images/logo_unand.png" width="200">

<br>

**FAKULTAS TEKNOLOGI INFORMASI**

**DEPARTEMEN INFORMATIKA**

**UNIVERSITAS ANDALAS**

**2026**

</div>

## PREFACE

Praise and gratitude be to Almighty God for all His grace and blessings so that this practicum report on Python Programming titled "Store Transaction Simulation and Access Validation System" could be completed successfully and on time.

This report was prepared to fulfill the practicum assignment requirements and to deepen our understanding of fundamental concepts in Python programming, specifically regarding the application of arithmetic, comparison, logical, assignment, membership, identity, and bitwise operators.

The author realizes that this report still has shortcomings. Therefore, constructive criticism and suggestions are highly expected for the future perfection of this report. May this report provide useful insights for all readers.

Padang, 2026

Laiba Shamshad

---
# BAB I
# PENDAHULUAN

## 1.1 Background

With the rapid evolution of information technology, automated systems are essential for processing transactions and controlling access privileges effectively. Manual verification processes are prone to calculation errors and operational bottlenecks. Utilizing basic programming structures allows for dynamic execution of pricing models, status authorizations, and promo evaluations.   

Python is an ideal language for developing complex logical pipelines due to its straightforward syntax and support for comprehensive operator categories. Arithmetic operators perform mathematical computations; comparison and logical operators validate structural requirements; membership and identity operators evaluate element presence and memory reference integrity; and bitwise operators manage low-level bitmask authorization codes efficiently.   

This practicum involves implementing six fundamental operator scripts and constructing the tugas3_8002.py integrated cashier program. The system handles raw data inputs, executes calculation rules, validates discount privileges, evaluates memory management logic, and constructs status bitmasks using student ID variable naming conventions (_8002).   


## 1.2 Objectives
1.Master the implementation of Python's standard operator suite: arithmetic, assignment, comparison, logical, bitwise, membership, and identity operators.   

2.Design an integrated console application (tugas3_8002.py) using custom variable naming conventions based on the student ID suffix (_8002).   

3.Evaluate binary status flags and customer access permissions using low-level bitwise operations (&, |, ^, <<).   

## 1.3 Benefits
1Provides practical fluency in dynamic data manipulation and mathematical formulation in Python.   

2.Establishes a foundational understanding of data validation techniques, memory identity checking, and binary bitmask manipulation for real-world application design.

---
# BAB II
# PEMBAHASAN

## 2.1 Basic theory

Arithmetic Operators: Perform mathematical calculations including addition (+), subtraction (-), multiplication (*), floating-point division (/), integer division (//), modulo (%), and exponentiation (**).   

Assignment Operators: Store values into variables. Augmented assignments combine arithmetic and assignment operations into single operations such as +=, -=, *=, /=, //=, %=, and **=.   

Comparison Operators: Evaluate relational equality between two operands (>, <, >=, <=, ==, !=), returning a boolean result (True or False).   

Logical Operators: Combine boolean expressions using conjunction (and), disjunction (or), and negation (not).   

Membership Operators: Determine element presence within sequence objects using in and not in.   

Identity Operators: Verify whether two variables share identical memory locations using is and is not, contrasting with value equality (==).   

Bitwise Operators: Perform bit-level operations on integer binary representations using AND (&), OR (|), XOR (^), NOT (~), left-shift (<<), and right-shift (>>).

## 2.2 Kode Program

### 2.2.1 ARITHMATIKA_2611528002
Run the code cell below so that Jupyter automatically saves the code along with its output as a single unit (no separate screenshot needed for the execution output).

In [1]:
#create a file name arithmetic_NIM.py
#create a program for arithmetic operators in python
# Variable names are appended with the last 4 digits of student nim
#THis program uses the input() function
#the entered value will be converted to an integer data type

number1 = int(input("input number - 1:  "))
number2 = int(input("input number - 2:  "))

#Addition
result = number1 + number2
print("\nAddition Operator")
print("Result =", result)

#Subtraction
result = number1 - number2
print("\nSubtraction Operator")
print("Result =", result)

#Multiplication
result = number1 * number2
print("\nMultiplication Operator")
print("Result =", result)

#Division integer division, and modulo
if number2 != 0:
    result = number1 / number2
    print("\nDivision Operator")
    print("Result =", result)
       
    result = number1 // number2
    print("\n Integer Division Operator")
    print("Result =", result)
              
    result = number1 % number2
    print("\nModulo Operator")
    print("Result =", result)
else:
    print(" The second number cannot be 0")
    result = number1 ** number2 
    print("\nExponentiation Operator")
    print("Result =", result)  
           






Addition Operator
Result = 34

Subtraction Operator
Result = -10

Multiplication Operator
Result = 264

Division Operator
Result = 0.5454545454545454

 Integer Division Operator
Result = 0

Modulo Operator
Result = 12


### Output Explanation:

1.Addition Operator (Result = 34): Calculates 12+22=34.   

2.Subtraction Operator (Result = -10): Calculates 12−22=−10.   

3.Multiplication Operator (Result = 264): Calculates 12×22=264.   

4.Division Operator (Result = 0.5454545454545454): Evaluates 12/22, performing standard floating-point division to return the full decimal quotient.   

5.Integer Division Operator (Result = 0): Evaluates 12//22, which discards the decimal fraction entirely and rounds down to the nearest whole integer (0).   

6.Modulo Operator (Result = 12): Evaluates 12%22, returning the remainder left over after integer division (12).   

Note on Exponentiation: Because number2 was 22 (not zero), the program executed the if block for division. The else block containing exponentiation (**) was skipped entirely.

### 2.2.2 ASSIGHMENT_2611528002"

In [2]:
# Create a file named assignment_NIM.py
# Variable names are appended with the last 4 digits of the student ID, example: number1_1234
# This program uses the input() function
# The entered value will be converted to an integer data type
# Assignment operator program in Python

number1 = int(input("Input number-1: "))
number2 = int(input("Input number-2: "))

print("\nInitial value of number1 =", number1)
print("Value of number2 =", number2)

# Standard assignment
result = number1
print("\nStandard assignment (=)")
print("Result =", result)

# Addition assignment
result = number1
result += number2
print("\nAddition assignment (+=)")
print("Result =", result)

# Subtraction assignment
result = number1
result -= number2
print("\nSubtraction assignment (-=)")
print("Result =", result)

# Multiplication assignment
result = number1
result *= number2
print("\nMultiplication assignment (*=)")
print("Result =", result)

# Division, integer division, and modulo assignment
if number2 != 0:
    result = number1
    result /= number2
    print("\nDivision assignment (/=)")
    print("Result =", result)
    # Additional operator
    result = number1
    result //= number2
    print("\nInteger division assignment (//=)")
    print("Result =", result)
    result = number1
    result %= number2
    print("\nModulo assignment (% =)")
    print("Result =", result)
else:
    print("\nDivision cannot be performed.")
    print("The second number cannot be 0.")

# Additional operator: exponentiation assignment
result = number1
result **= number2
print("\nExponentiation assignment (**=)")
print("Result =", result)


Initial value of number1 = 9
Value of number2 = 12

Standard assignment (=)
Result = 9

Addition assignment (+=)
Result = 21

Subtraction assignment (-=)
Result = -3

Multiplication assignment (*=)
Result = 108

Division assignment (/=)
Result = 0.75

Integer division assignment (//=)
Result = 0

Modulo assignment (% =)
Result = 9

Exponentiation assignment (**=)
Result = 282429536481


### OUTPUT EXPLANATION 

When executing the program with inputs `number1 = 9` and `number2 = 12`:
1. Initial Display: The program displays the starting values of `number1` (9) and `number2` (12).
2. Standard Assignment (`=`): Assigns the value `9` to `result`.
3. Addition Assignment (`+=`): Resets `result` to `9`, adds `12`, and assigns `21` to `result`.
4. Subtraction Assignment (`-=`): Resets `result` to `9`, subtracts `12`, yielding `-3`.
5. Multiplication Assignment (`*=`): Resets `result` to `9`, multiplies by `12`, yielding `108`.
6. Division Assignment (`/=`): Resets `result` to `9`, divides by `12` to produce floating-point `0.75`.
7. Integer Division Assignment (`//=`): Resets `result` to `9`, performs floor division by `12`, rounding down to `0`.
8. Modulo Assignment (`%=`): Resets `result` to `9`, evaluates $9 \pmod{12}$, yielding the remainder `9`.
9. Exponentiation Assignment (`**=`): Resets `result` to `9`, raises it to the power of `12` ($9^{12}$), yielding `282429536481`.


## 2.2.3 BITWISE_2611528002

In [3]:
# Create a file named bitwise_NIM.py
# Variable names are appended with the last 4 digits of the student ID (example: angka1_1234)
# This program uses the input() function

print("\n==============================")
print("3. BITWISE OPERATOR")
print("==============================")

angka1 = int(input("Enter bitwise number-1: "))
angka2 = int(input("Enter bitwise number-2: "))

print("\nNumbers in decimal and binary form")
print("angka1 =", angka1, "| binary =", bin(angka1))
print("angka2 =", angka2, "| binary =", bin(angka2))

# Bitwise AND
hasil = angka1 & angka2
print("\nBitwise AND (&)")
print(angka1, "&", angka2, "=", hasil)
print("Binary result =", bin(hasil))
print("Binary result (8 bit) =", format(hasil, "08b"))

# Bitwise OR
hasil = angka1 | angka2
print("\nBitwise OR (|)")
print(angka1, "|", angka2, "=", hasil)
print("Binary result =", bin(hasil))
print("Binary result (8 bit) =", format(hasil, "08b"))

# Bitwise XOR
hasil = angka1 ^ angka2
print("\nBitwise XOR (^)")
print(angka1, "^", angka2, "=", hasil)
print("Binary result =", bin(hasil))
print("Binary result (8 bit) =", format(hasil, "08b"))

# Bitwise NOT
hasil = ~angka1
print("\nBitwise NOT (~)")
print("~", angka1, "=", hasil)
print("Binary result =", bin(hasil))
print("Binary result (8 bit) =", format(hasil, "08b"))

# Bitwise left shift
jumlah_geser = int(input("\nEnter the number of bit shifts: "))

hasil = angka1 << jumlah_geser
print("\nBitwise left shift (<<)")
print(angka1, "<<", jumlah_geser, "=", hasil)
print("Binary result =", bin(hasil))
print("Binary result (8 bit) =", format(hasil, "08b"))

# Bitwise right shift
hasil = angka1 >> jumlah_geser
print("\nBitwise right shift (>>)")
print(angka1, ">>", jumlah_geser, "=", hasil)
print("Binary result =", bin(hasil))
print("Binary result (8 bit) =", format(hasil, "08b"))


3. BITWISE OPERATOR

Numbers in decimal and binary form
angka1 = 33 | binary = 0b100001
angka2 = 24 | binary = 0b11000

Bitwise AND (&)
33 & 24 = 0
Binary result = 0b0
Binary result (8 bit) = 00000000

Bitwise OR (|)
33 | 24 = 57
Binary result = 0b111001
Binary result (8 bit) = 00111001

Bitwise XOR (^)
33 ^ 24 = 57
Binary result = 0b111001
Binary result (8 bit) = 00111001

Bitwise NOT (~)
~ 33 = -34
Binary result = -0b100010
Binary result (8 bit) = -0100010

Bitwise left shift (<<)
33 << 11 = 67584
Binary result = 0b10000100000000000
Binary result (8 bit) = 10000100000000000

Bitwise right shift (>>)
33 >> 11 = 0
Binary result = 0b0
Binary result (8 bit) = 00000000


### OUTPUT EXPLANATION:

When executing the program with inputs `angka1 = 33`, `angka2 = 24`, and shift count `11`:
1. Binary Representation: The program converts `33` to binary `00100001` and `24` to `00011000`.
2. Bitwise AND (`&`): Compares corresponding bits and sets to `1` only if both are `1`. Since no positions overlap, the result is `0` (`00000000`).
3. Bitwise OR (`|`): Combines active bits from both numbers, resulting in `00111001` (`57` in decimal).
4. Bitwise XOR (`^`): Sets bits to `1` where values differ. With non-overlapping inputs, this matches the OR operation, returning `57`.
5. Bitwise NOT (`~`): Negates all bits using standard two's complement syntax `~x = -(x + 1)`, turning `33` into `-34`.
6. Bitwise Left Shift (`<<`): Moves bits 11 positions to the left (multiplying by $2^{11}$), resulting in `67584`.
7. Bitwise Right Shift (`>>`): Moves bits 11 positions to the right (performing floor division by $2^{11}$), reducing the value to `0`.


## 2.2.4 LAINNYA_2611528002

In [4]:
# Create a file named lainnya_NIM.py (others_NIM.py)
# Variable names are appended with the last 4 digits of the student ID, example: number1_1234            # This program uses the input() function
# Membership and identity operator program

print("================================")
print("1. MEMBERSHIP OPERATOR")
print("================================")

# Input several data separated by commas
input_data = input("Enter several numbers, separate with commas: ")

# Convert input into an integer list
data = [int(number.strip()) for number in input_data.split(",")]

search_value = int(input("Enter the number you want to search for: "))

# Operator in
result = search_value in data
print("\nMembership operator IN")
print(search_value, "in", data, "=", result)

# Operator not in
result = search_value not in data
print("\nMembership operator NOT IN")
print(search_value, "not in", data, "=", result)

print("\n================================")
print("2. IDENTITY OPERATOR")
print("================================")

# objek1 uses the list from user input
objek1 = data

# objek2 refers to the same object as objek1
objek2 = objek1

# objek3 has the same contents, but is a new object
objek3 = data.copy()

print("objek1 =", objek1)
print("objek2 =", objek2)
print("objek3 =", objek3)

# Operator is
result = objek1 is objek2
print("\nIdentity operator IS")
print("objek1 is objek2 =", result)

# Operator is not
result = objek1 is not objek3
print("\nIdentity operator IS NOT")
print("objek1 is not objek3 =", result)

# Comparing identity and value
print("\nComparison of identity and value")
print("objek1 is objek3 =", objek1 is objek3)
print("objek1 == objek3 =", objek1 == objek3)

1. MEMBERSHIP OPERATOR

Membership operator IN
4 in [2, 3, 6, 12, 32, 11] = False

Membership operator NOT IN
4 not in [2, 3, 6, 12, 32, 11] = True

2. IDENTITY OPERATOR
objek1 = [2, 3, 6, 12, 32, 11]
objek2 = [2, 3, 6, 12, 32, 11]
objek3 = [2, 3, 6, 12, 32, 11]

Identity operator IS
objek1 is objek2 = True

Identity operator IS NOT
objek1 is not objek3 = True

Comparison of identity and value
objek1 is objek3 = False
objek1 == objek3 = True


## OUTPUT EXPLANATION

When executing the program with list `[2, 3, 6, 12, 32, 11]` and search target `4`:
1. Membership `in`: Evaluates `4 in [2, 3, 6, 12, 32, 11]` and returns `False` because `4` is not an element of the sequence.
2. Membership `not in`: Evaluates `4 not in [2, 3, 6, 12, 32, 11]` and returns `True` confirming the number is absent.
3. Identity `is`: Compares `objek1` and `objek2`. Because `objek2` points directly to the memory reference of `objek1`, `objek1 is objek2` yields `True`.
4. Identity `is not`: Compares `objek1` and `objek3`. Because `objek3` was instantiated as a shallow copy (`data.copy()`), it occupies a separate memory location, so `objek1 is not objek3` returns `True`.
5. Identity vs Value Comparison: 
   - `objek1 is objek3` evaluates to `False` because identity requires matching memory addresses.
   - `objek1 == objek3` evaluates to `True` because standard equality only verifies that the list contents and sequence are identical.

## 2.2.4 LOGIC_2611528002

In [5]:
  # Create a file named logic_NIM.py
  # Variable names are appended with the last 4 digits of the student ID, example: a1_1234
  # This program uses the input() function
  # Logical operator program in Python

  # Entering boolean values
  # Input is not case-sensitive (for uppercase and lowercase letters)
a1 = input("Input boolean value-1 (true/false): ").strip().lower() == "true"
a2 = input("Input boolean value-2 (true/false): ").strip().lower() == "true"

print("\nA1 =", a1)
print("A2 =", a2)

 # Conjunction: evaluates to True if both are True
result = a1 and a2
print("\nConjunction (AND)")
print("A1 and A2 =", result)

 # Disjunction: evaluates to True if at least one is True
result = a1 or a2
print("\nDisjunction (OR)")
print("A1 or A2 =", result)

 # Negation A1: reverses the value of A1
result = not a1
print("\nNegation A1 (NOT)")
print("not A1 =", result)

 # Negation A2: reverses the value of A2
result = not a2
print("\nNegation A2 (NOT)")
print("not A2 =", result)

 # XOR: evaluates to True if both values are different
result = a1 != a2
print("\nExclusive Disjunction (XOR)")
print("A1 XOR A2 =", result)


A1 = False
A2 = False

Conjunction (AND)
A1 and A2 = False

Disjunction (OR)
A1 or A2 = False

Negation A1 (NOT)
not A1 = True

Negation A2 (NOT)
not A2 = True

Exclusive Disjunction (XOR)
A1 XOR A2 = False


## OUTPUT EXPLANATION

When executing the program where both inputs parse as `a1 = False` and `a2 = False`:
1. **Input Parsing**: The user inputs are sanitized with `.strip().lower()` and compared against `"true"`. Since neither input matched, both `a1` and `a2` store `False`.
2. **Conjunction (`AND`)**: Evaluates `False and False` to `False` because the AND operation requires all terms to be `True`.
3. **Disjunction (`OR`)**: Evaluates `False or False` to `False` because neither term satisfies the minimum condition of being `True`.
4. **Negation A1 (`NOT`)**: Inverts `a1` (`not False`) to return `True`.
5. **Negation A2 (`NOT`)**: Inverts `a2` (`not False`) to return `True`.
6. **Exclusive Disjunction (`XOR`)**: Evaluates `a1 != a2` (`False != False`). Since both operands share identical boolean states, the inequality check returns `False`.

## 2.2.5 PERBANDIGAN_2611528002

In [6]:
# Create a file named perbandigan/comparison_NIM.py
# Variable names are appended with the last 4 digits of the student ID, example: number1_1234
# This program uses the input() function
# The entered value will be converted to an integer data type
# Comparison operator program in Python

number1 = int(input("Input number-1: "))
number2 = int(input("Input number-2: "))

# Greater than
result = number1 > number2
print("\nGreater than operator")
print("number1 > number2 =", result)

# Less than
result = number1 < number2
print("\nLess than operator")
print("number1 < number2 =", result)

# Greater than or equal to
result = number1 >= number2
print("\nGreater than or equal to operator")
print("number1 >= number2 =", result)

# Less than or equal to
result = number1 <= number2
print("\nLess than or equal to operator")
print("number1 <= number2 =", result)

# Equal to
result = number1 == number2
print("\nEqual to operator")
print("number1 == number2 =", result)

# Not equal to
result = number1 != number2
print("\nNot equal to operator")
print("number1 != number2 =", result)

# Additional: chained comparison in Python
result = 0 < number1 < 100
print("\nChained comparison")
print("0 < number1 < 100 =", result)

result = 0 < number2 < 100
print("0 < number2 < 100 =", result)


Greater than operator
number1 > number2 = False

Less than operator
number1 < number2 = True

Greater than or equal to operator
number1 >= number2 = False

Less than or equal to operator
number1 <= number2 = True

Equal to operator
number1 == number2 = False

Not equal to operator
number1 != number2 = True

Chained comparison
0 < number1 < 100 = True
0 < number2 < 100 = True


## OUTPUT EXPLANATION

When executing the program with inputs `number1 = 12` and `number2 = 22`:
1. Greater Than (`>`): Evaluates `12 > 22` and returns `False`.
2. Less Than (`<`): Evaluates `12 < 22` and returns `True`.
3. Greater Than or Equal To (`>=`): Evaluates `12 >= 22` and returns `False`.
4. Less Than or Equal To (`<=`): Evaluates `12 <= 22` and returns `True`.
5. Equal To (`==`): Evaluates `12 == 22` and returns `False` because the two inputs are distinct.
6. Not Equal To (`!=`): Evaluates `12 != 22` and returns `True`.
7. Chained Comparison (`0 < x < 100`): Evaluates whether each value falls strictly within the open interval $(0, 100)$. Both `12` and `22` satisfy the compound relational condition, yielding `True` for both lines.

---
## Tugas Praktikum

### Store Transaction Simulation and Access Validation System

You are asked to create a simple cashier program used to calculate a customer's total transaction while also checking the customer's status and access rights to certain promotions. The program receives transaction data and customer information, then uses various types of Python operators to determine the total payment, discount status, promo eligibility, and customer access rights. The entire process must be built into a single, complete program flow.


**Kode Jawaban :**

a.  Kode Program "Alpro Tugas3_B_2611528002"

In [7]:
# assignment3_2611528002.py
# Week 3 Assignment - Algorithms and Programming Practicum 2026

# --- 1. USER INPUT & DATA TYPES ---
name_1234 = input("Enter Customer Name: ")
status_1234 = input("Enter Customer Status (member/non-member): ").strip().lower()
total_1234 = float(input("Enter Total Purchase: "))
item_count_1234 = int(input("Enter Number of Items: "))
promo_code_1234 = input("Enter Promo Code: ").strip().upper()

# List of available promo codes
promo_list_1234 = ["HEMAT10", "HEMAT20", "GRATISONGKIR"]

# --- 2. MEMBERSHIP OPERATORS ---
# Check if the promo code exists in the list
valid_promo_1234 = promo_code_1234 in promo_list_1234        # 'in' operator
invalid_promo_1234 = promo_code_1234 not in promo_list_1234  # 'not in' operator

# --- 3. COMPARISON OPERATORS ---
# Evaluates expressions into boolean True or False values
min_purchase_met_1234 = total_1234 >= 200000  # '>=' operator
min_items_met_1234 = item_count_1234 >= 3      # '>=' operator
is_member_1234 = status_1234 == "member"      # '==' operator

# --- 4. LOGICAL OPERATORS ---
# Combines multiple conditions using and, or, and not
gets_discount_1234 = is_member_1234 and min_purchase_met_1234  # 'and' operator
gets_promo_1234 = (min_items_met_1234 or valid_promo_1234) and not (total_1234 < 50000)  # 'or', 'not'

# --- 5. ARITHMETIC OPERATORS ---
# Calculate discount amount, final payment, average item price, and remainder
discount_rate_1234 = 0.10 if gets_discount_1234 else 0.0
discount_amount_1234 = total_1234 * discount_rate_1234  # '*' multiplication operator
final_payment_1234 = total_1234 - discount_amount_1234  # '-' subtraction operator
avg_item_price_1234 = total_1234 / item_count_1234      # '/' division operator
item_remainder_1234 = item_count_1234 % 2              # '%' modulus operator

# --- 6. ASSIGNMENT OPERATORS ---
# Calculate loyalty points using augmented assignment
loyalty_points_1234 = 0
loyalty_points_1234 += int(final_payment_1234 // 10000)  # '+=' and '//' floor division operators

# --- 7. IDENTITY OPERATORS ---
# Demonstrating object identity ('is', 'is not') vs value equality ('==')
obj_a_1234 = [1, 2, 3]
obj_b_1234 = [1, 2, 3]
obj_c_1234 = obj_a_1234

same_identity_1234 = (obj_a_1234 is obj_c_1234)      # 'is' operator (True)
diff_identity_1234 = (obj_a_1234 is not obj_b_1234)  # 'is not' operator (True)

# --- 8. BITWISE OPERATORS ---
# Construct Bitmask based on customer metrics:
# Bit 0 (0001 = 1): Member status
# Bit 1 (0010 = 2): Total purchase >= Rp200,000
# Bit 2 (0100 = 4): Item count >= 3
# Bit 3 (1000 = 8): Valid promo code

bit_member_1234 = 1 if is_member_1234 else 0
bit_purchase_1234 = 2 if min_purchase_met_1234 else 0
bit_items_1234 = 4 if min_items_met_1234 else 0
bit_promo_1234 = 8 if valid_promo_1234 else 0

# Bitwise OR (|) to combine into a single status code
status_code_1234 = bit_member_1234 | bit_purchase_1234 | bit_items_1234 | bit_promo_1234

# Bitwise AND (&) to verify individual conditions
check_member_1234 = status_code_1234 & 1
check_promo_1234 = status_code_1234 & 8

# Bitwise XOR (^) to compare against reference baseline (1011 = 11)
reference_code_1234 = 11
xor_result_1234 = status_code_1234 ^ reference_code_1234

# Bitwise Left Shift (<<)
shifted_code_1234 = status_code_1234 << 1

# --- 9. DISPLAY PROGRAM OUTPUT ---
print("\n=== STORE TRANSACTION SYSTEM ===")
print("=== TRANSACTION DATA ===")
print(f"Customer Name: {name_1234}")
print(f"Customer Status: {status_1234}")
print(f"Total Purchase: Rp{total_1234:.0f}")
print(f"Number of Items: {item_count_1234}")
print(f"Promo Code: {promo_code_1234}")

print("\n=== VALIDATION RESULTS ===")
print(f"Purchase >= Rp200000: {min_purchase_met_1234}")
print(f"Number of Items >= 3: {min_items_met_1234}")
print(f"Member Status: {is_member_1234}")
print(f"Promo Code Available: {valid_promo_1234}")
print(f"Gets Discount: {gets_discount_1234}")
print(f"Gets Promo: {gets_promo_1234}")

print("\n=== CALCULATION RESULTS ===")
print(f"Discount: Rp{discount_amount_1234:.0f}")
print(f"Total Payment: Rp{final_payment_1234:.0f}")
print(f"Average Item Price: Rp{avg_item_price_1234:.2f}")
print(f"Loyalty Points Earned: {loyalty_points_1234}")

print("\n=== IDENTITY OPERATOR DEMO ===")
print(f"obj_a == obj_b (Value Check): {obj_a_1234 == obj_b_1234}")
print(f"obj_a is obj_b (Identity Check): {obj_a_1234 is obj_b_1234}")
print(f"obj_a is obj_c (Identity Check): {same_identity_1234}")

print("\n=== BITWISE OPERATIONS ===")
print("=== Transaction Status Code ===")
print(f"Binary Code: {status_code_1234:04b}")
print(f"Decimal Code: {status_code_1234}")

print("\n=== Status Check ===")
print(f"Check Member ({status_code_1234:04b} & 0001): {check_member_1234:04b} (Dec: {check_member_1234})")
print(f"Check Promo  ({status_code_1234:04b} & 1000): {check_promo_1234:04b} (Dec: {check_promo_1234})")

print("\n=== Status Comparison ===")
print(f"Transaction Code: {status_code_1234:04b}")
print(f"Reference Code  : {reference_code_1234:04b}")
print(f"XOR Result ({status_code_1234:04b} ^ {reference_code_1234:04b}): {xor_result_1234:04b} (Dec: {xor_result_1234})")

print("\n=== Shift ===")
print(f"Left Shift ({status_code_1234:04b} << 1): {shifted_code_1234:05b} (Dec: {shifted_code_1234})")
print("\n=== DONE ===")


=== STORE TRANSACTION SYSTEM ===
=== TRANSACTION DATA ===
Customer Name: laiba
Customer Status: member
Total Purchase: Rp15
Number of Items: 14
Promo Code: 0024

=== VALIDATION RESULTS ===
Purchase >= Rp200000: False
Number of Items >= 3: True
Member Status: True
Promo Code Available: False
Gets Discount: False
Gets Promo: False

=== CALCULATION RESULTS ===
Discount: Rp0
Total Payment: Rp15
Average Item Price: Rp1.07
Loyalty Points Earned: 0

=== IDENTITY OPERATOR DEMO ===
obj_a == obj_b (Value Check): True
obj_a is obj_b (Identity Check): False
obj_a is obj_c (Identity Check): True

=== BITWISE OPERATIONS ===
=== Transaction Status Code ===
Binary Code: 0101
Decimal Code: 5

=== Status Check ===
Check Member (0101 & 0001): 0001 (Dec: 1)
Check Promo  (0101 & 1000): 0000 (Dec: 0)

=== Status Comparison ===
Transaction Code: 0101
Reference Code  : 1011
XOR Result (0101 ^ 1011): 1110 (Dec: 14)

=== Shift ===
Left Shift (0101 << 1): 01010 (Dec: 10)

=== DONE ===


## OUTPUT EXPLANATION:

When executing the integrated cashier script with inputs `name = "laiba"`, `status = "member"`, `total = 15`, `item_count = 14`, and `promo = "0024"`:

1. Validation & Logical Operations:
   - `min_purchase_met`: Evaluates `15 >= 200000` to `False`.
   - `min_items_met`: Evaluates `14 >= 3` to `True`.
   - `is_member`: Matches `"member"` to return `True`.
   - `valid_promo`: Checks membership of `"0024"` in the promo list, returning `False`.
   - `gets_discount`: Combines `is_member` and `min_purchase_met` via `AND`, returning `False`.
   - `gets_promo`: Evaluates `(True or False) and not (15 < 50000)`. Because `not (True)` becomes `False`, the entire expression evaluates to `False`.

2. Calculations:
   - `Discount`: Set to `0` because no discount was earned.
   - `Total Payment`: Remains `Rp15`.
   - `Average Item Price`: Divides spend by quantity ($15 / 14$), yielding `Rp1.07`.
   - `Loyalty Points`: Evaluates floor division $15 // 10000$, yielding `0`.

3. Identity vs Equality:
   - `obj_a == obj_b` is `True` because list values match.
   - `obj_a is obj_b` is `False` because the lists occupy separate memory slots.
   - `obj_a is obj_c` is `True` because both point to the exact same reference.

4. Bitwise Bitmask Processing:
   - Active flags (Member = `0001`, Items = `0100`) combine via Bitwise OR into bitmask `0101` (Decimal `5`).
   - Bitwise AND confirms member status (`0101 & 0001 = 0001`) and invalid promo status (`0101 & 1000 = 0000`).
   - Bitwise XOR compares `0101` against reference `1011`, returning difference code `1110` (Decimal `14`).
   - Bitwise Left Shift (`0101 << 1`) shifts bits left by 1 position, doubling decimal value `5` to `10` (`01010`).

## Analysis of Assignment Questions

Q1. Arithmetic Operators Usage in Transaction Logic

. Multiplication (*): Determines the raw discount deduction by multiplying the total gross transaction by the percentage rate.   

. Subtraction (-): Computes the final net payment by subtracting the total discount from the gross total.   

. Division (/): Evaluates the unit item price average across total items purchased.   

. Modulo (%): Identifies leftover integer remainders when total spend is divided by quantity units.   

Q2. Distinction Between Logical and, or, and not

. and: Returns True only if all expressions evaluate to True (e.g., verifying is_member AND total spend ≥Rp200,000).   

. or: Returns True if at least one sub-condition evaluates to True (e.g., qualifying for promos via spending OR buying ≥3 items).   

. not: Inverts boolean results, turning True to False and vice versa.   

Q3. Purpose of Membership Operators (in / not in)

. in: Scans the promo list ["HEMAT10", "HEMAT20", "GRATISONGKIR"] and yields True if the user's promo code matches any list item.   

. not in: Returns True if the input string is absent from the valid list.   

Q4. Identity (is) vs Equality (==) Comparison

. == (Value Equality): Evaluates whether two objects share identical data values regardless of memory storage address.   

. is (Identity Comparison): Verifies whether two references point directly to the exact same memory address. Separate objects containing identical elements yield True under == but False under is.   

Q5. Customer Access Privilege Evaluation using Bitwise Operators
 Bitwise operators compress complex state flags into binary positions (0001 Member, 0010 Spend, 0100 Quantity, 1000 Promo):   

. Bitwise OR (|): Combines binary flags into an integrated bitmask access integer.   

. Bitwise AND (&): Isolates specific bit positions to evaluate specific privileges.   

. Bitwise XOR (^): Identifies structural differences between actual status bitmasks and baseline permissions.   

Q6. Complete Execution Sequence

. Data Ingestion: Collect customer attributes using dynamic input prompts.   

. Validation Stage: Execute membership and comparison checks against set criteria.   

. Financial Calculation: Calculate discount deductions and update total payments using assignment operators (-=).   

. Bitmask Processing: Form binary access status codes via bitwise combination operators.   

. Console Output Generation: Output structured results to the screen.   


Q7. Specific Program Boolean Outcomes

. True Outcome: min_pembelian_met_8002 evaluates to True when input purchase is 250000 (250000≥200000).   

. False Outcome: gets_discount_8002 evaluates to False when customer status is "non-member", even if total purchase exceeds Rp200,000.   

---
#CHAPTER III

CONCLUSION

Based on the completed practicum exercises, we conclude that:

1.Python operators (arithmetic, assignment, comparison, logical, membership, identity, and bitwise) provide a framework for handling both simple data calculations and complex decision systems.   

2.Augmented assignments streamline state updates, logical and relational operators enforce validation rules, and bitwise bitmasks offer a flexible method for storing multiple status flags in an integrated numeric variable.

---
REFERENCES

[1] Python Software Foundation, "Python 3 Documentation: Expressions & Built-in Types," [Online]. Available: https://docs.python.org/3/reference/expressions.html. [Accessed: 18 September 2026].   


[2] Practicum Lecturer Team, Algorithms and Programming Practicum Guide 2026: Week 3 - Python Operators, Department of Informatics, Universitas Andalas, 2026.   
